format/applebookmark/apple_bookmark.go

```diff
@@ -0,0 +1,311 @@
+	})
+
+	return hdr
+}
+
+type tocEntry struct {
+	key          uint64
+	recordOffset int64
+}
+
+const (
+	arrayEntrySize = 4
+	dictEntrySize  = 4
+)
+
+func decodeRecord(d *decode.D) {
+	d.FieldStruct("record", func(d *decode.D) {
+		n := int(d.FieldU32("length"))
+		typ := d.FieldU32("type", dataTypeMap)
+		switch typ {
+		case dataTypeString:
+			d.FieldUTF8("data", n)
+		case dataTypeData:
+			d.FieldRawLen("data", int64(n*8))
+		case dataTypeNumber8:
+			d.FieldS8("data")
+		case dataTypeNumber16:
+			d.FieldS16("data")
+		case dataTypeNumber32:
+			d.FieldS32("data")
+		case dataTypeNumber64:
+			d.FieldS64("data")
+		case dataTypeNumber32F:
+			d.FieldF32("data")
+		case dataTypeNumber64F:
+			d.FieldF64("data")
+		case dataTypeDate:
+			d.FieldF64BE("data", scalar.DescriptionTimeFn(scalar.S.TryActualF, cocoaTimeEpochDate, time.RFC3339))
+		case dataTypeBooleanFalse:
+		case dataTypeBooleanTrue:
+		case dataTypeArray:
+			d.FieldStructNArray("data", "element", int64(n/arrayEntrySize), func(d *decode.D) {
+				offset := calcOffset(d.FieldU32("offset"))
+				d.SeekAbs(offset, decodeRecord)
+			})
+		case dataTypeDictionary:
+			d.FieldStructNArray("data", "element", int64(n/dictEntrySize), func(d *decode.D) {
+				keyOffset := calcOffset(d.FieldU32("key_offset"))
+				d.FieldStruct("key", func(d *decode.D) {
+					d.SeekAbs(keyOffset, decodeRecord)
+				})
+
+				valueOffset := calcOffset(d.FieldU32("value_offset"))
+				d.FieldStruct("value", func(d *decode.D) {
+					d.SeekAbs(valueOffset, decodeRecord)
+				})
+			})
+		case dataTypeUUID:
+			d.FieldRawLen("data", int64(n*8))
+		case dataTypeURL:
+			d.FieldUTF8("data", n)
+		case dataTypeRelativeURL:
+			baseOffset := d.FieldU32("base_url_offset")
+			d.FieldStruct("base_url", func(d *decode.D) {
+				d.SeekAbs(int64(baseOffset), decodeRecord)
+			})
+
+			suffixOffset := d.FieldU32("suffix_offset")
+			d.FieldStruct("suffix", func(d *decode.D) {
+				d.SeekAbs(int64(suffixOffset), decodeRecord)
+			})
+		}
+	})
+}
+
+const reservedSize = 32
+const headerEnd = 48
+const headerEndBitPos = headerEnd * 8
+
+// all offsets are calculated relative to the end of the bookmark header
+func calcOffset(i uint64) int64 { return int64(8 * (i + headerEnd)) }
+
+func bookmarkDecode(d *decode.D, _ any) any {
+
+	// all fields are little-endian with the exception of the Date datatype.
+	d.Endian = decode.LittleEndian
+
+	// decode bookmarkdata header, one at the top of each "file",
+	// although these may be nested inside of binary plists
+	d.FieldStruct("header", func(d *decode.D) {
+		d.FieldUTF8("magic", 4, d.AssertStr("book", "alis"))
+		d.FieldU32("total_size")
+		d.FieldU32("unknown")
+		d.FieldU32("header_size", d.AssertU(48))
+		d.FieldRawLen("reserved", reservedSize*8)
+	})
+
+	tocOffset := calcOffset(d.FieldU32("first_toc_offset"))
+
+	var currentHdr *tocHeader
+	var tocHeaders []*tocHeader
+
+	d.FieldArrayLoop("toc_headers", func() bool {
+		return tocOffset != headerEndBitPos
```

Comment:
Some limit on this loop? btw you can do `make fuzz GROUP=apple_bookmark` to fuzz just one format, have been a very efficient way to find infinite loop and excessive memory use

Reply:
Nice, `make fuzz` caught a case of infinite recursion in the `decodeRecord` function which makes sense. Added a check for that and set what I think is a reasonable limit of 100 TOCs here as well.